handler/src/main/java/io/netty/handler/ssl/ReleasableOpenSslEngine.java

```diff
@@ -0,0 +1,1988 @@
+        final int bioRead;
+
+        if (dst.isDirect() && dst.remaining() >= pending) {
+            final int pos = dst.position();
+            final long addr = Buffer.address(dst) + pos;
+            bioRead = SSL.readFromBIO(networkBIO, addr, pending);
+            if (bioRead > 0) {
+                dst.position(pos + bioRead);
+                return bioRead;
+            }
+        } else {
+            final ByteBuf buf = alloc.directBuffer(pending);
+            try {
+                final long addr = memoryAddress(buf);
+
+                bioRead = SSL.readFromBIO(networkBIO, addr, pending);
+                if (bioRead > 0) {
+                    int oldLimit = dst.limit();
+                    dst.limit(dst.position() + bioRead);
+                    buf.getBytes(0, dst);
+                    dst.limit(oldLimit);
+                    return bioRead;
+                }
+            } finally {
+                buf.release();
+            }
+        }
+
+        return bioRead;
+    }
+
+    private SSLEngineResult readPendingBytesFromBIO(ByteBuffer dst, int bytesConsumed, int bytesProduced,
+                                                    SSLEngineResult.HandshakeStatus status) throws SSLException {
+        // Check to see if the engine wrote data into the network BIO
+        int pendingNet = SSL.pendingWrittenBytesInBIO(networkBIO);
+        if (pendingNet > 0) {
+
+            // Do we have enough room in dst to write encrypted data?
+            int capacity = dst.remaining();
+            if (capacity < pendingNet) {
+                return new SSLEngineResult(BUFFER_OVERFLOW,
+                        mayFinishHandshake(status != FINISHED ? getHandshakeStatus(pendingNet) : status),
+                        bytesConsumed, bytesProduced);
+            }
+
+            // Write the pending data from the network BIO into the dst buffer
+            int produced = readEncryptedData(dst, pendingNet);
+
+            if (produced <= 0) {
+                // We ignore BIO_* errors here as we use in memory BIO anyway and will do another SSL_* call later
+                // on in which we will produce an exception in case of an error
+                SSL.clearError();
+            } else {
+                bytesProduced += produced;
+                pendingNet -= produced;
+            }
+            // If isOuboundDone is set, then the data from the network BIO
+            // was the close_notify message -- we are not required to wait
+            // for the receipt the peer's close_notify message -- shutdown.
+            if (isOutboundDone) {
+                shutdown();
+            }
+
+            return new SSLEngineResult(getEngineStatus(),
+                    mayFinishHandshake(status != FINISHED ? getHandshakeStatus(pendingNet) : status),
+                    bytesConsumed, bytesProduced);
+        }
+        return null;
+    }
+
+    @Override
+    public final SSLEngineResult wrap(
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "wrap" is 41 which is greater than 25 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)